hal/src/nRF52840/usart_hal.cpp

```diff
@@ -226,49 +228,76 @@ class Usart {
         enableTimer();
         startReceiver();
 
-        return 0;
+        return SYSTEM_ERROR_NONE;
     }
 
     int end() {
         CHECK_TRUE(enabled_, SYSTEM_ERROR_INVALID_STATE);
         AtomicSection lk;
 
-        disableInterrupts();
-
-        stopTransmission();
-        stopReceiver();
-
-        nrf_uarte_disable(uarte_);
+        suspend_ = false;
 
-        nrf_uarte_txrx_pins_disconnect(uarte_);
-        nrf_uarte_hwfc_pins_disconnect(uarte_);
+        CHECK(disable());
 
-        HAL_Pin_Mode(txPin_, INPUT);
+        // Configuring the input pins' mode to PIN_MODE_NONE will disconnect input buffer to enable power savings
+        // Configuring the output pins' mode to INPUT_PULLUP to not polluting the state on other side.
+        HAL_Pin_Mode(txPin_, INPUT_PULLUP);
         HAL_Set_Pin_Function(txPin_, PF_NONE);
-        HAL_Pin_Mode(rxPin_, INPUT);
+        HAL_Pin_Mode(rxPin_, PIN_MODE_NONE);
         HAL_Set_Pin_Function(rxPin_, PF_NONE);
-
         if (config_.config & SERIAL_FLOW_CONTROL_CTS) {
-            HAL_Pin_Mode(ctsPin_, INPUT);
+            HAL_Pin_Mode(ctsPin_, PIN_MODE_NONE);
             HAL_Set_Pin_Function(ctsPin_, PF_NONE);
         }
         if (config_.config & SERIAL_FLOW_CONTROL_RTS) {
-            HAL_Pin_Mode(rtsPin_, INPUT);
+            HAL_Pin_Mode(rtsPin_, INPUT_PULLUP);
             HAL_Set_Pin_Function(rtsPin_, PF_NONE);
         }
 
-        nrfx_prs_release(uarte_);
+        enabled_ = false;
+        transmitting_ = false;
+        receiving_ = 0;
+        config_ = {};
+        rxBuffer_.reset();
+        txBuffer_.reset();
 
-        disableTimer();
+        return SYSTEM_ERROR_NONE;
+    }
+
+    int suspend() {
```

Comment:
In order for things to be somewhat more robust, if using hardware flow control we should probably deassert RTS first to make sure that the other side does not send anything while we deinitialize things. Same applies to STM32 implementation.

Reply:
I don't see this implemented. This will probably work fine on Gen 3 (again, probably). Both on Gen 2 and Gen 3 we should be able to disconnect HWFC pins from the peripheral and manually assert the necessary state.